gateway-messages/src/lib.rs

```diff
@@ -90,6 +93,57 @@ pub enum SpPort {
     Two = 2,
 }
 
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum UpdateStatus {
+    /// The SP has no update status.
+    None,
+    /// Returned when the SP is still preparing to apply the update with the
+    /// given ID (e.g., erasing a target flash slot).
+    Preparing(UpdatePreparationStatus),
+    /// Returned when an update is currently in progress.
+    InProgress(UpdateInProgressStatus),
+    /// Returned when an update has completed.
+    ///
+    /// The SP has no concept of time, so we cannot indicate how recently this
+    /// update completed. The SP will continue to return this status until a new
+    /// update starts (or the status is reset some other way, such as an SP
+    /// reboot).
+    Complete(UpdateId),
+    /// Returned when an update has been aborted.
+    ///
+    /// The SP has no concept of time, so we cannot indicate how recently this
+    /// abort happened. The SP will continue to return this status until a new
+    /// update starts (or the status is reset some other way, such as an SP
+    /// reboot).
+    Aborted(UpdateId),
+}
+
+/// Current state when the SP is preparing to apply an update.
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub struct UpdatePreparationStatus {
+    pub id: UpdateId,
+    pub progress: Option<UpdatePreparationProgress>,
+}
+
+/// Current progress of preparing for an update.
+///
+/// The initial values reported by the SP should have `current=0` and `total`
+/// defined in some SP-specific unit. `current` should advance toward `total`;
+/// once `current == total` preparation is complete, and the SP should return
+/// `UpdateStatus::InProgress` instead.
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub struct UpdatePreparationProgress {
```

Comment:
Not sure, but I sort of feel like I don't want to do that? For preparation, currently the units for host boot flash are "sectors erased", but I could imagine that being different for other components ("pages erased" maybe? I'm not sure what other prep steps might be besides erasing, though.)

What if I combine this with @ahl's comment above, and inline the progress of update delivery but leave this intentionally unitless?

```rust
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, JsonSchema)]
#[serde(tag = "state", rename_all = "snake_case")]
enum SpUpdateStatus {
    /// The SP has no update status.
    None,
    /// The SP is preparing to receive an update.
    ///
    /// May or may not include progress, depending on the capabilities of the
    /// component being updated.
    Preparing { id: Uuid, progress: Option<UpdatePreparationProgress> },
    /// The SP is currently receiving an update.
    InProgress {
        id: Uuid,
        bytes_received: u32,
        total_bytes: u32,
    },
    /// The SP has completed receiving an update.
    Complete { id: Uuid },
    /// The SP has aborted an in-progress update.
    Aborted { id: Uuid },
}

/// Progress of an SP preparing to update.
///
/// The units of `current` and `total` are unspecified and defined by the SP;
/// e.g., if preparing for an update requires erasing a flash device, this may
/// indicate progress of that erasure without defining units (bytes, pages,
/// sectors, etc.). 
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, JsonSchema)]
struct UpdatePreparationProgress {
    current: u32,
    total: u32,
}

Reply:
> What if I combine this with @ahl's comment above, and inline the progress of update delivery but leave this intentionally unitless?

Works for me!

Reply:
Applied in 4b09d42f